### ISSUE TITLE
FSDP: issue with cpu offloading & async

### ISSUE THREAD

**Issue by min-xu-ai** [facebookresearch/fairscale#636]
@myleott, can you add more details on this issue you have run into? Thanks a lot!

**Comment by myleott** [comment]
Ideally we would use non-blocking GPU -> CPU transfer when doing CPU offloading. This was originally enabled, but I had to disable it in #496 because I found it was producing non-identical results.

I was not able to build a unit test to reproduce this problem, but it happens with the fairseq example here: https://github.com/pytorch/fairseq/tree/master/examples/fully_sharded_data_parallel#13b-params-on-1-v100-gpu-with-cpu-offloading

Basically, you should see identical results with `non_blocking=True` and `non_blocking=False`, but we don't. This suggests a missing stream synchronization or something similar.
